open_spiel/games/2048.cc

```diff
@@ -0,0 +1,383 @@
+// Copyright 2022 DeepMind Technologies Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "open_spiel/games/2048.h"
+
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
+#include "open_spiel/spiel_utils.h"
+#include "open_spiel/utils/tensor_view.h"
+
+namespace open_spiel {
+namespace two_zero_four_eight {
+namespace {
+
+constexpr int kMoveUp = 0;
+constexpr int kMoveRight = 1;
+constexpr int kMoveDown = 2;
+constexpr int kMoveLeft = 3;
+const std::vector<Action> kPlayerActions
+   = {kMoveUp, kMoveRight, kMoveDown, kMoveLeft};
+
+// Facts about the game.
+const GameType kGameType{/*short_name=*/"2048",
+                         /*long_name=*/"2048",
+                         GameType::Dynamics::kSequential,
+                         GameType::ChanceMode::kExplicitStochastic,
+                         GameType::Information::kPerfectInformation,
+                         GameType::Utility::kGeneralSum,
+                         GameType::RewardModel::kTerminal,
+                         /*max_num_players=*/1,
+                         /*min_num_players=*/1,
+                         /*provides_information_state_string=*/false,
+                         /*provides_information_state_tensor=*/false,
+                         /*provides_observation_string=*/true,
+                         /*provides_observation_tensor=*/true};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new TwoZeroFourEightGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+}  // namespace
+
+TwoZeroFourEightState::TwoZeroFourEightState(std::shared_ptr<const Game> game)
+    : State(game) {
+  board_ = std::vector<Tile>(kDefaultRows * kDefaultColumns, Tile(0, false));
+}
+
+void TwoZeroFourEightState::SetCustomBoard(const std::vector<int> board_seq) {
+  current_player_ = 0;
+  for (int x = 0; x < kDefaultRows; x++) {
+    for (int y = 0; y < kDefaultColumns; y++) {
+      SetBoard(x, y, Tile(board_seq[x * kDefaultRows + y], false));
+    }
```

Comment:
use `r` and `c` rather than `x` and `y` as better names

Reply:
Have changed this